### ISSUE TITLE
Need help building it for the PlayStation Classic Mini (Armv7)

### ISSUE THREAD

**Issue by shoddydev** [fgsfdsfgs/perfect_dark#403]
Hello, I'm trying to build it for the PlayStation Classic which is a linux armv7 device that came out in Dec 2018. I'm using gcc 12 and SDL2 2.0.12. I've successfully compiled it but when I try to run it it gives this error:

ERROR: FATAL: Could not open SDL window
failed to create an EGL window surface

![IMG_9360](https://github.com/fgsfdsfgs/perfect_dark/assets/42045893/d2c3d01b-6574-4236-a2be-7d1b60b6016e)

I also tried building it against SDL2 2.0.14 and I see this warning:
Assertion failture at SDL_GetWindowDisplayIndex_REAL (SDL2-2.0.14/src/video/SDL_Video.c:1080), triggered 1 time:
   'window && window->magic == &_this->window_magic'

![IMG_9362](https://github.com/fgsfdsfgs/perfect_dark/assets/42045893/8e51ae74-8a07-489d-9026-6016919180a6)

Then I tried editing gfx_sdl2.cpp to delete this
    if (!wnd) {
        sysFatalError("Could not open SDL window:\n%s", SDL_GetError());
    }

and it would give me more clues:
![IMG_9363](https://github.com/fgsfdsfgs/perfect_dark/assets/42045893/91f612ef-894b-4a21-99ce-cad8cee5d332)

it would say Invalid window so I'm assuming that it's not able to connect to a window? I've compiled a few PC Ports previously for the PSC, Zelda soh, sonic mania, gta3, and I would really like to get Perfect Dark working.

The PSC uses wayland and weston 1.11.0, and OpenGL ES 3.2 which btw, does perfect dark have GLES support? I'm no expert I'm really just a amateur but initially I came across a similar issue when building zelda soh port. But luckily someone had just added GLES support and it worked.

I'm compiling it on something called psclassic Debian which is a linux debian distro for the PSC. Interestingly enough, the build worked on there. However it's running through software. I am told by the author of the distro that it's wrapped in software, so OpenGL is emulated which is why it runs super slow. It should run full speed if running it on the stock os. here's a video of it:
https://www.youtube.com/watch?v=pPyXUKRvBas

**Comment by fgsfdsfgs** [comment]
There's no GLES support at the moment, though theoretically it wouldn't be very hard to add. Looks like the game tries to request a desktop GL context and then fails, presumably because it's not supported. Not sure about that assert, that'd be something related to the SDL port for that particular platform, or it might be related to SDL's Wayland backend.  
Can you run it with `--debug-gl --log` and then post the log (`pd.log`) here?  
If the platform is strictly ES only, you'd have to to use a wrapper like gl4es for now.

**Comment by shoddydev** [comment]
I forgot to mention that I compiled gl4es to get libGL.so.1. here's the logfile:
[pd.log](https://github.com/fgsfdsfgs/perfect_dark/files/14608594/pd.log)

it just says
version: port c9f880e42 (armv7l-linux)
startup date: 01 Sep 2018 14:59:20
base dir: ./data
save dir: /media/Apps/perfectdark/home/.local/share/perfectdark
ERROR: FATAL: Could not open SDL window:
failed to create an EGL window surface

Ah okay i'm still confused about what GLES is. Normally I just try a project without checking if it supports it or not. I assume SDL2 handle the rendering. I should check that before I try new projects. I wouldn't know where to begin when it comes to adding GLES support. if you can lead me to the right direction I'll be very grateful. but first I should probably figure out the SDL window problem.

**Comment by fgsfdsfgs** [comment]
GLES = OpenGL ES is essentially the mobile/embedded device variant of OpenGL. It generally has a smaller feature set, but GLES 3.2 is probably enough to cover all features required by the game.

To support it, two main things have to be done: the shader generator has to be modified to generate GLES3.2-compliant shaders, and the game should be able to request a GLES context instead of a normal GL context when creating the window.

SDL is not used for the actual 3D graphics part, it is used to interact with your window system/display server to get a window, request a GL context for it and then the actual rendering is done using GL.

**Comment by shoddydev** [comment]
I've been told by seasoned programmers not to use ChatGPT but it actually helped me in adding GLES support.  it's useful for clueless beginners like me.
https://www.youtube.com/watch?v=8d6GthLnKRU

I probably butchered the code though, but it works. basically I edited 3 files, gfx_opengl.cpp, gfx_sdl2.cpp, and Makefile.port. i'll clone this project and add the changes later

**Comment by shoddydev** [comment]
I'm seeing a graphical issue, there's like a layer priority problem where the gun goes behind other objects when it should be in front:
![graphic1](https://github.com/fgsfdsfgs/perfect_dark/assets/42045893/e3e356e3-fa1d-4f5a-975c-024f88b0988f)

https://www.youtube.com/watch?v=xfZVvDcwUR0

Any idea what can cause this? I assume it's something in the gfx_opengl.cpp file since I changed that to add GLES support.
[gfx_opengl.cpp.txt](https://github.com/user-attachments/files/16059400/gfx_opengl.cpp.txt)

**Comment by fgsfdsfgs** [comment]
Without looking at the diff in detail, if I had to guess, you probably removed the call to `glDepthRange`.
UPD:
You're still loading GL with glad, which if you haven't regenerated `glad.c` will still attempt to load desktop GL3.2. This might cause you to mix desktop GL and GLES calls, which might be the problem. In particular, `glDepthRange` in GLES is called `glDepthRangef`. This also applies to some other functions like `glClearDepthf`.

**Comment by shoddydev** [comment]
I should probably mention that i'm using a commit from march 14th so the code might be dated. I did try to compile the latest but whenever I enter make -f Makefile.port to update some code changes, it would re-compile from scratch. the old one would just update changes so i went with that as it was faster to build.

and yeah that's what i get for trying to use ChatGPT as a guide lol well it got me further than I would have if I tried to do it on my own. I'm really just a hobbyist so I'm clueless on what these calls are. I'll remove glad.c and tinker around some more.

**Comment by shoddydev** [comment]
So I think I found the problem though I'm not entirely sure. I was able to edit the gfx_opengl.cpp to remove glad and all the GL calls and made it GLES only. However the problem with the graphic layer still persist. So I did a lot of tinkering and I stumbled on the env.c file. 

basically at the struct nofogenvironment g_NoFogEnvironments[] = { area I see a section with all the stages. 
This here is the first solo mission stage

{ STAGE_DEFECTION,     10, 10000,   0,  0,  0, RGB(0x000000), NO_SUNS,            0, RGB(0x3a1100),  5000,   0, 0, RGB(0x000000),      0,    0, 0, 1 },

I changed the 10 value to 1 and it worked. First I tried setting it higher and the hand gun layer would sink further behind all the other layers. 2 works good too but I notice the handgun still goes behind the enemy partially. Is setting it to 1 okay? And is there anything I should look out for when doing this.

**Comment by shoddydev** [comment]
Any idea what causes this black rectangle to appear in the sky? I see it various stages that has a great distance in the background.
![001](https://github.com/user-attachments/assets/12d8ea71-4a2b-423e-8953-47ea1efe3339)

I found a way to decrease it by editing the env.c file. here i edited the STAGE_AIRBASE "far" value from 20000 to 32767. 
![002b](https://github.com/user-attachments/assets/2a2eff76-7830-4646-9290-aae5808922fb)

I notice the box gets smaller the higher i increase the value.
![003](https://github.com/user-attachments/assets/cc14e1bc-a532-4483-9bc0-1d19c687666a)

 If I go higher than 32767, then it looks like this
![004](https://github.com/user-attachments/assets/844359da-50f4-4694-a203-21ce9a36fd55)

basically the gun layer gets sent back again
![005](https://github.com/user-attachments/assets/f034c5dd-5ab9-4774-bd9b-38106a49344f)

The black rectangle becomes more noticeable in cut scenes like this
https://www.youtube.com/live/sgxlZbFcoII?si=YRr7fRVN-Ac5nAqh&t=550

**Comment by fgsfdsfgs** [comment]
That's the far clipping plane. Either the framebuffer is not cleared with the correct color or something is wrong with the fog or depth. Like I said before, depth-related functions and some other stuff has to be adapted for GL ES. I'll get around to it eventually, I hope. This can be done without editing the game part of the code.

**Comment by shoddydev** [comment]
> That's the far clipping plane. Either the framebuffer is not cleared with the correct color or something is wrong with the fog or depth. Like I said before, depth-related functions and some other stuff has to be adapted for GL ES. I'll get around to it eventually, I hope. This can be done without editing the game part of the code.

with the help of ChatGPT I actually removed glad and got it to compile with only GL ES. here's my gfx_opengl.cpp file
[gfx_opengl.cpp.txt](https://github.com/user-attachments/files/17823685/gfx_opengl.cpp.txt)

**Comment by shoddydev** [comment]
I'm looking at actual N64 footage compared to mines and it's pretty different. Looks like the water on mines is scaled up more
![001](https://github.com/user-attachments/assets/66d93fc1-b1ae-44aa-9cde-60be50186a74)
![002](https://github.com/user-attachments/assets/0451855a-bf55-49d0-b463-5c422f5dcf85)
![003](https://github.com/user-attachments/assets/3d0141d2-0109-4ecf-871b-3b7e9a8cce24)
![004](https://github.com/user-attachments/assets/df909263-e33b-4261-bd36-ddfdb0f65f3f)

**Comment by fgsfdsfgs** [comment]
I pushed some workarounds for GLES3 support. Try them out, see if it fixes the depth.

**Comment by shoddydev** [comment]
> I pushed some workarounds for GLES3 support. Try them out, see if it fixes the depth.

well I should mention that I'm using a commit from April, i tried the newer ones but it wouldn't init for some reason and I noticed the latest one uses cmake. I'm compiling directly on a PSC using make and getting cmake to work on a PSC is not easy. I also tried cross compiling on debian but when I try to run it it didn't init so that's why I built it directly on a PSC using this sm64_builder app i put together.

Anyway I tried just updating the files that were changed recently and I got a ton of errors. But good news, I just edited one line in my original gles code and it fixed the issue. 

basically i noticed this section in the updated code
        // HACK: workaround for no GL_DEPTH_CLAMP
        append_line(vs_buf, &vs_len, "    gl_Position.z *= 0.3f;");

And I added that line in my code like this:
append_line(vs_buf, &vs_len, "gl_Position = aVtxPos;");
append_line(vs_buf, &vs_len, "    gl_Position.z *= 0.3f;");
append_line(vs_buf, &vs_len, "}");

and it fixed the issue. I even reverted the env.c file back to original and now everything looks correct and behaving like it should. Thanks again man I really appreciate your help
